Move most methods off Variable into torch::autograd::impl functions.

ezyang opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29907 Add more defensive tests for undefined tensors.
* #28287 Merge Tensor and Variable types.
* #29667 Delete redefinitions of methods in Variable already present on Tensor.
* **#29665 Move most methods off Variable into torch::autograd::impl functions.**
* #29653 Remove TensorImpl::is_variable, deprecate Tensor::is_variable

Our intention is to merge the static distinction between Tensor and
Variable.  Ordinarily, this would entail merging the methods of Tensor
and Variable.  But there are a lot of "private"-ish methods on Variable
that we don't actually want to dump onto the Tensor class.  So, as prep
work, we move all of those methods off of Variable and into
the torch::autograd::impl namespace (impl as in, please don't use this
end users).  This ends up being a fairly large patch because all of
the call sites have to play ball too.

While I was on the topic, I also moved any of the touched functions into
the C++ file, so that modifying them would not trigger a recompilation of
all of torch.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18496169](https://our.internmc.facebook.com/intern/diff/D18496169)

ezyang committed:
Move most methods off Variable into torch::autograd::impl functions.

Our intention is to merge the static distinction between Tensor and
Variable.  Ordinarily, this would entail merging the methods of Tensor
and Variable.  But there are a lot of "private"-ish methods on Variable
that we don't actually want to dump onto the Tensor class.  So, as prep
work, we move all of those methods off of Variable and into
the torch::autograd::impl namespace (impl as in, please don't use this
end users).  This ends up being a fairly large patch because all of
the call sites have to play ball too.

While I was on the topic, I also moved any of the touched functions into
the C++ file, so that modifying them would not trigger a recompilation of
all of torch.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Move most methods off Variable into torch::autograd::impl functions."


Our intention is to merge the static distinction between Tensor and
Variable.  Ordinarily, this would entail merging the methods of Tensor
and Variable.  But there are a lot of "private"-ish methods on Variable
that we don't actually want to dump onto the Tensor class.  So, as prep
work, we move all of those methods off of Variable and into
the torch::autograd::impl namespace (impl as in, please don't use this
end users).  This ends up being a fairly large patch because all of
the call sites have to play ball too.

While I was on the topic, I also moved any of the touched functions into
the C++ file, so that modifying them would not trigger a recompilation of
all of torch.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

ezyang committed:
Update on "Move most methods off Variable into torch::autograd::impl functions."


Our intention is to merge the static distinction between Tensor and
Variable.  Ordinarily, this would entail merging the methods of Tensor
and Variable.  But there are a lot of "private"-ish methods on Variable
that we don't actually want to dump onto the Tensor class.  So, as prep
work, we move all of those methods off of Variable and into
the torch::autograd::impl namespace (impl as in, please don't use this
end users).  This ends up being a fairly large patch because all of
the call sites have to play ball too.

While I was on the topic, I also moved any of the touched functions into
the C++ file, so that modifying them would not trigger a recompilation of
all of torch.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

[ghstack-poisoned]

kostmo commented:
## CircleCI build failures summary
As of commit 1a08c82:
* **1/1** broken upstream at merge base 6d54c5d (see [grid view](https://dr.pytorch.org/master-timeline.html?count=10&sha1=6d54c5ddd22ea69f6adaacc428d6c60aa63c8b77&should_suppress_scheduled_builds=true&should_suppress_fully_successful_columns=true&max_columns_suppress_successful=35))
    * You may want to rebase on the `viable/strict` branch (see [age history](https://dr.pytorch.org/master-viable-commits.html)):
      ```
      git fetch viable/strict
      git rebase viable/strict
      ```
* **0/1** failures introduced in this PR
* **0/1** recognized as flaky



Here are the [reasons each build failed](https://dr.pytorch.org/commit-details.html?sha1=1a08c82cbcaadb12347c84a3c38e9569c942e0cc):

| Job | Step | Log excerpt |
| --- | --- | --- |
| pytorch_xla_linux_xenial_py3_6_clang7_build | Build | `Failed to run '['/var/lib/jenkins/workspace/xla/scripts/generate_code.sh']'` |


---
This comment was automatically generated by [Dr. CI](https://github.com/kostmo/circleci-failure-tracker/tree/master/docs/from-pull-request-comment).
Follow [this link to opt-out](https://dr.pytorch.org/admin/comments-opt-out.html) of these comments for your Pull Requests.


Please report bugs/suggestions on the [GitHub issue tracker](https://github.com/kostmo/circleci-failure-tracker/issues).

ezyang committed:
Update on "Move most methods off Variable into torch::autograd::impl functions."


Our intention is to merge the static distinction between Tensor and
Variable.  Ordinarily, this would entail merging the methods of Tensor
and Variable.  But there are a lot of "private"-ish methods on Variable
that we don't actually want to dump onto the Tensor class.  So, as prep
work, we move all of those methods off of Variable and into
the torch::autograd::impl namespace (impl as in, please don't use this
end users).  This ends up being a fairly large patch because all of
the call sites have to play ball too.

While I was on the topic, I also moved any of the touched functions into
the C++ file, so that modifying them would not trigger a recompilation of
all of torch.

Signed-off-by: Edward Z. Yang <ezyang@fb.com>

Differential Revision: [D18496169](https://our.internmc.facebook.com/intern/diff/D18496169)

[ghstack-poisoned]

albanD reviewed (commented):
You didn't changed the moved functions (appart from arguments massaging) right?

ezyang commented on torch/csrc/autograd/variable.cpp:
```diff
@@ -151,53 +126,25 @@ const std::shared_ptr<Node>& Variable::grad_fn() const {
   (*list)[pos] = nullptr;
 }
 
+uint32_t Variable::current_version() const noexcept {
```
Oops, forgot to rename this.

ezyang replied:
Oh it gets renamed next PR

ezyang commented:
> You didn't changed the moved functions (appart from arguments massaging) right?

Yep, no changes.

ezyang commented on torch/csrc/autograd/variable.h:
```diff
@@ -757,53 +669,17 @@ inline bool Variable::is_view() const noexcept {
     : at::Tensor(std::move(self)) {}
 
-inline at::TensorImpl* Variable::get() const {
-  TORCH_CHECK(defined(), "Called Variable::get() on an undefined Variable");
```
It's probably safer to add definedness checks to all of the sites that used to call get()

ezyang replied:
Actually, all of the implementations that get moved to VariableType are safe because if you have an undefined tensor you won't dispatch here in the first place.

ezyang replied:
For the other ones, I stacked it as a diff on top as incorporating into this diff would have resulted in some merge conflicts. https://github.com/pytorch/pytorch/pull/29907

ezyang commented on torch/csrc/autograd/variable.h:
```diff
@@ -289,27 +294,7 @@ struct TORCH_API Variable : public at::Tensor {
   //~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
 
-  void set_name(const std::string& name);
   const std::string& name() const noexcept;
```
I kept name because `x.name` is supported in Python. Setting of this name is not supported so I removed the setter.

ezyang commented on torch/csrc/autograd/variable.h:
```diff
@@ -276,6 +278,9 @@ struct TORCH_API Variable : public at::Tensor {
   hook_return_var_t<T> register_hook(T&& hook);
 
+  // Remove hook at given position
+  void remove_hook(unsigned pos);
```
`register_hook` is public API, and it returns an index, to delete you have to call `remove_hook`. @yf225's suggestion is that we fix this API to return a remove handle, but for now the returned index doesn't make much sense unless you also have `remove_hook` on the API.
